File: computing_methods/CourseWork/src/GradientMethod.scala
class GradientMethod(iterationsLimits: Int) extends LinearSystemsSolvingMethod {
  override def solve(s: LinearEquationSystem, epsilon: Double): (Vector, Int) = {
    val A = s.getA
    val b = s.getB
    val at = A.transpose
    var x = new Vector(Array.tabulate[Double](b.getDimension)(i => -b(i) / A(i, i)))
    var rp = A * x - b
    var iterations = 0
    while (rp.norm > epsilon) {
      val w = A * at * rp
      val mu = rp * w / (w.norm * w.norm)
      x = x - (at * rp * mu)
      iterations = iterations + 1
      if (iterations > iterationsLimits || x.hasNan) return null
      rp = A * x - b
    }
    (x, iterations)
  }
}

File: computing_methods/SystemOfLinearEquations/src/GradientMethod.java
import javafx.util.Pair;

public class GradientMethod implements SolvingMethod {
    public final int iterationsLimit;

    public GradientMethod(int iterationsLimit) {
        this.iterationsLimit = iterationsLimit;
    }

    @Override
    public Pair<Vector, Integer> solve(LinearEquationsSystem system, double epsilon) {
        Matrix A = system.getA();
        Vector b = system.getB();
        Matrix at = A.transpose();
        Vector x = new Vector(b.getDimensions());
        for (int i = 0; i < b.getDimensions(); i++) x.set(i, -b.get(i) / A.get(i, i));
        Vector rp = A.multiply(x).subtract(b);
        int iterations = 0;
        while (rp.norm() > epsilon) {
            Vector w = A.multiply(at).multiply(rp);
            double mu = rp.multiply(w) / (w.norm() * w.norm());
            x = x.subtract(A.multiply(rp).multiply(mu));
            iterations++;
            if (iterations > iterationsLimit || x.hasNaN())
                return null;
            rp = A.multiply(x).subtract(b);
        }
        return new Pair<>(x, iterations);
    }
}
